File: src/main/java/com/jcs/suadeome/generators/IdGenerator.java
package com.jcs.suadeome.generators;

import com.jcs.suadeome.types.Id;

import java.util.function.Supplier;

import static com.jcs.suadeome.types.Id.id;

public class IdGenerator {

  private final Supplier<String> value;

  public IdGenerator(Supplier<String> value) {
    this.value = value;
  }

  public Id generate(EntityType type) {
    return id(type.getIdPrefix() + value.get());
  }
}

File: src/main/kotlin/com/jcs/suadeome/generators/IdGenerator.kt
package com.jcs.suadeome.generators

import com.jcs.suadeome.types.Id
import java.time.format.DateTimeFormatter

class IdGenerator(private val clock: Clock, val random: Random) {

    fun generate(type: EntityType): Id {
        val timestamp = clock.utcNow().format( DateTimeFormatter.ofPattern("yyDDDHHmmssSSS") )
        val random = random.random(100..999)
        return Id(type.idPrefix + timestamp + random)
    }

    companion object {
        fun default(): IdGenerator = IdGenerator(RealClock(), RealRandom() )
    }


}




